### PR TITLE
Milightv6: Spring clean up

### LINKED ISSUE #2918: [milight] Milight floods the inbox
Every milight bridge deliberately floods the inbox with things of all _possible_ thing types, not only those which _really_ exist (see [README.md](https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.milight/README.md)). I understand that the protocol doesn't support any proper discovery for its things. Then, however, the binding should also not support it instead of doing this.

### LINKED ISSUE #3046: [Milight] Multiple Keep-alive timers after some time
After installing openhab 2.2, milight flood the logs only with V6 bridge (v3 doesn't do that).
```
15:32:43.991 [INFO ] [smarthome.event.ThingUpdatedEvent    ] - Thing 'milight:bridgeV6:PontMilight2' has been updated.
15:32:44.284 [INFO ] [smarthome.event.ThingUpdatedEvent    ] - Thing 'milight:bridgeV6:PontMilight2' has been updated
```
and events : 
```
event: message
data: {"topic":"smarthome/things/milight:bridgeV6:PontMilight2/updated","payload":"[{\"label\":\"iBox/iBox2\",\"configuration\":{\"REFRESH_IN_SEC\":5,\"REPEAT\":1,\"WAIT_BETWEEN_COMMANDS\":100,\"ID\":\"0000000012BD\",\"CUSTOM_PORT\":3390,\"PASSWORD_BYTE_1\":0,\"ADDR\":\"192.168.2.15\",\"PASSWORD_BYTE_2\":0},\"properties\":{\"sessionid\":\"00 00\",\"sessionid_last_refresh\":\"1514557579255\"},\"UID\":\"milight:bridgeV6:PontMilight2\",\"thingTypeUID\":\"milight:bridgeV6\",\"channels\":[]},{\"label\":\"iBox/iBox2\",\"configuration\":{\"REFRESH_IN_SEC\":5,\"REPEAT\":1,\"WAIT_BETWEEN_COMMANDS\":100,\"ID\":\"0000000012BD\",\"CUSTOM_PORT\":3390,\"PASSWORD_BYTE_1\":0,\"ADDR\":\"192.168.2.15\",\"PASSWORD_BYTE_2\":0},\"properties\":{\"sessionid\":\"00 00\",\"sessionid_last_refresh\":\"1514557579255\"},\"UID\":\"milight:bridgeV6:PontMilight2\",\"thingTypeUID\":\"milight:bridgeV6\",\"channels\":[]}]","type":"ThingUpdatedEvent"}
```
the bug is reported on the community forum : 
https://community.openhab.org/t/the-new-milight-ibox-and-milight-binding/17621/279

### LINKED ISSUE #3069: [MiLight] Please add INCREASE/DECREASE for CHANNEL_COLOR
In Basic UI when you long press arrows on a colorpicker widget, it send INCREASE or DECREASE command eclipse/smarthome#3009
But in actual milight binding, for COLOR_CHANNEL (the one used for Colorpicker item), INCREASE or DECREASE aren't supported https://github.com/openhab/openhab2-addons/blob/d4bee788dd84f8728bb6c8df8fd63c2663fbda3a/addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/handler/MilightLedHandler.java#L99
It will be find to handle this message and change bulb's brightness

### LINKED ISSUE #3237: [Milight] [2.3] 2.4 Ghz RW 4 Zone Controller not working
Mi.Light RGBW LED Strip 2.4GHz RF Wireless 4-Zone Controller Receiver Box is not working with latest Openhab Milight Binding (2.3.0) . It worked with 2.2 without problems.

## Expected Behavior
Should work as it did in openhab 2.2

## Current Behavior
No reaction from my led strip ... with the milight android app its working. 

The config is the same as it was with openhab2 2.2

```
Bridge milight:bridgeV6:F0FE6B2DBE90 [ ADDR="192.168.1.47", ID="F0FE6B2DBE90" ] {    
    Thing rgbwLed 1
}
```

## Possible Solution
None

## Steps to Reproduce (for Bugs)
Try to control the 4 Zone Controller

## Context
?

## Your Environment
* Version used: 2.3.0
* Environment name and version: irrelevant
* Operating System and version: ubuntu 16

### LINKED ISSUE #3398: [Milight] Night mode issues (flash on activate, won't come out of night mode)
With night mode, the light will flash on at what ever it's brightness was, if the light was off before going into night mode. In addition, it sometimes won't come out of night mode (power on command needs to be sent before it'll respond to any other commands).

  I've actually fixed these in my own fork @ https://github.com/xnhil/openhab2-addons/tree/2.2.x_milight But I'm new to GitHub and OH development so I'm not sure of the process for getting this incorporated/reviewed.

  Any feedback welcome :)

### LINKED ISSUE #3798: Milight Gateway V3 Discovery Packet Error Wrong IP Address in Log
## Current Behavior

I am having Milight Wifi Gateway/Bridge v3, I am getting Continous Error logs in Openhab logs.

My IP address for Gateway/Bridge is 192.168.1.135, but in logs continous it reflects wrong ip address.

> 2018-08-04 21:34:48.963 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V3 discovery packet to 192.168.1.136 failed. Socket is closed
> 
> 2018-08-04 21:34:48.965 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V6 discovery packet to 192.168.1.136 failed. Socket is closed
> 
> 2018-08-04 21:34:49.167 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V3 discovery packet to 192.168.1.136 failed. Socket is closed
> 
> 2018-08-04 21:34:49.169 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V6 discovery packet to 192.168.1.136 failed. Socket is closed
> 
> 2018-08-04 21:34:49.370 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V3 discovery packet to 192.168.1.136 failed. Socket is closed
> 
> 2018-08-04 21:34:49.372 [ERROR] [ht.internal.protocol.MilightDiscover] - Sending a V6 discovery packet to 192.168.1.136 failed. Socket is closed

image

## Steps to Reproduce (for Bugs)

1. I aleady tried to uninstall Binding and Re-install, but no effects.
2.

## Context

How to get rid off this issue ? my logs is crazy filled with all Errors

## Your Environment

* Version used: (e.g., openhab and addon versions)
* Environment name and version (e.g. Chrome 39, node.js 5.4, Java 8, ...):
* Operating System and version (desktop or mobile, Windows 10, Raspbian Jessie, ...):
Openhab Version 2.2
Milight Binding

### LINKED ISSUE #4064: [milight] RF collisions causing sporadic responses from globes
[milight] binding not accounting for RF collisions causing sporadic responses from multiple bridges. My testing environment has 2 ibox2 devices which work perfectly if you restrict commands to 1 of the 2 devices. The moment both bridges are included in an action the outcome become random. It may fully function or some bulbs may not respond.
When sending data to multiple bridges there needs to be a delay between sending packets to each bridge otherwise I believe RF collisions may occur (i.e. both bridges transmit data and the packets collide cause the globe to miss the packet).
Whilst there is a delay between re-transmits this appears to be accounting for delaying in transmitting to the same bridge. There appears to be no consideration for multiple bridges that are within range of each other (which in my place seems to be anywhere).
As what I've described will not have anything in logs of any value as programmatically there is no 'visible' error I'm not sure on the point of providing logs. I will note I've tried to resolve this myself using a dirty file creation (file used as locking mechanism) in one of the routines but I don't believe it's the best way. Therefore my logs probably aren’t overly helpful. If however someone feels default logs will in some way help just let me know and I’ll go back to the standard code and provide logs. 
## Possible Solution
I was simply trying to add some sort of thread lock or global 'transmission in progress flag' but it's somewhat beyond my programming skills.
## Steps to Reproduce (for Bugs)
Simply setup 2 bridges with a few zones each. Send commands to zones across 2 bridges at the same time. In my case in roughly 10 - 40% of cases the expected outcome (e.g. all lights on) will not fully occur (typically 1 or 2 lights do not respond). Send to only 1 bridge but all 4 zones (individually, not via zone 0) and I get a 100% success rate. 
## Context
Based on the lighting being a bit unreliable full deployment to the rest of the house isn't really an option.
## Your Environment
Openhab 2.3 (on PI3)
Milight binding 2.4.0
Currently testing with 2 ibox2 devices. Full deployment will be 15+ devices for full house individual light control.
Previously using about 7 ibox2 devices with my own nodeJS script to control lights with zero retransmits but serial communication (i.e. no parallel communication) to all bridges with 100% success rate. i.e. RF environment and globes are just fine.

### PR DESCRIPTION
**Pull request by davidgraeff**
Config: Breaking change.

* Make explicit handlers out of all the bulb types.
* Use configuration holder classes instead of manually parsing.
* Don't handle config changes ourselves, just use the default behaviour of dispose()/initialize().
* Use lower case config names as per convention (BREAKING).
* Use a zone config for bulbs and not the (generated) Thing ID as before (BREAKING).
* Don't add all available bulbs to the discovery. The user need to add his/her bulb manually.
  The protocol doesn't support bulb detection, we shouldn't pretend to either.
* Don't start another scheduler timer every few seconds for the V6 bridge session keepalive.
* Don't use the scheduler for the V3 bridge keep alive: We "abuse" the socket timeout time instead
  and make the code simpler.
* Use \<context> in thing-desc.xml
* Use global send queue for the entire binding. If multiple bridges are setup and they receive
  their commands in parallel that causes radio interference as reported by a user.
* Add increase/decrease support for channel color

Fixes #4064
Fixes #3798
Fixes #3398
Fixes #3237
Fixes #3069
Fixes #3046
Fixes #2918

Signed-off-by: David Graeff <david.graeff@web.de>

### CODE REVIEW

**Comment by davidgraeff** [comment]
@martinvw Some binding fixes. Tested with the old and new bridge with the new bridge being kindly donated by [Daan_Lindeman](https://community.openhab.org/u/Daan_Lindeman).

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/AbstractLedHandler.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+                    }
+                } else {
+                    logger.error("Relative animation mode channel only supports IncreaseDecreaseType");
+                }
+                break;
+            }
+            default:
+                logger.error("Channel unknown {}", channelUID.getId());
+        }
+    }
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/AbstractLedHandler.java]: That is a huge method?

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV3Handler.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                                "Bridge did not respond or the bridge's MAC address does not match with your configuration!");
+                        timeoutsCounter = 0;
+                    } else {
+                        socket.setSoTimeout(300);
+                        ++timeoutsCounter;
+                        socket.send(discoverPacketV3);
+                    }
+                    continue;
+                } catch (IOException e) {
+                    break; // Thread interrupted
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV3Handler.java]: the comment does imho not 1-on-1 match the caught exception?

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV6Handler.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * The {@link BridgeV6Handler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public class BridgeV6Handler extends AbstractBridgeHandler implements ISessionState {
+    @NonNullByDefault({})
+    private MilightV6SessionManager session;
+    final SimpleDateFormat sdf = new SimpleDateFormat("HH:mm");
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV6Handler.java]: Please note that SimpleDateFormat is not thread-safe, in java 8 [DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) which is both thread-safe and immutable.

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV6Handler.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    public MilightV6SessionManager getSessionManager() {
+        return session;
+    }
+
+    @Override
+    public void sessionStateChanged(SessionState state, @Nullable InetAddress newAddress) {
+        switch (state) {
+            case SESSION_VALID_KEEP_ALIVE:
+                preventReinit = true;
+                long t = session.getLastSessionValidConfirmation();
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV6Handler.java]: `t`?

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/AbstractLedHandler.java (hunk 1 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/AbstractLedHandler.java]: This was huge before, many other methods got smaller. I know that `AbstractLightHandler` deserves some slimming, but I don't feel like I wanna do it in this PR.

File: addons/binding/org.openhab.binding.milight/README.md [hunk 5]
```diff
@@ -39,19 +38,19 @@ configuration file.
 iBox and iBox2 have the version 6, older milight bridges have the version 3.
 The ID is the MAC address of the bridge in hexadecimal digits.
 
-    Bridge milight:bridgeV3:ACCF23A6C0B4 [ ADDR="192.168.0.70", ID="ACCF23A6C0B4" ] {
-        Thing whiteLed 0
-        Thing rgbwwLed 4
-        Thing rgbLed 1
+    Bridge milight:bridgeV3:mybridge [ host="192.168.0.70", bridgeid="ACCF23A6C0B4", passwordByte1=0, passwordByte2=0 ] {
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.milight/README.md]: I can't think of a way. I need two 0-255 restricted int values. With a string configuration the user can mess up a lot, and the code will grow to sort those wrong inputs.

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/discovery/MilightDiscover.java [hunk 6]
```diff
@@ -40,68 +44,44 @@
...
+    private final static boolean enableV6 = true;
 
     ///// Network
-    private byte[] discoverBufferV3 = "Link_Wi-Fi".getBytes();
-    private byte[] discoverBufferV6 = "HF-A11ASSISTHREAD".getBytes();
+    private final int receivePort;
     private final DatagramPacket discoverPacketV3;
     private final DatagramPacket discoverPacketV6;
     private boolean willbeclosed = false;
+    @NonNullByDefault({})
```
[the remark this answers is not in the document]
Comment (wborn) [reply addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/discovery/MilightDiscover.java]: It should always be added before the type declaration see [Null Annotations](https://www.openhab.org/docs/developer/development/conventions.html#null-annotations). 

Only when a variable has no modifiers the formatter will format the null annotation above the line.

**Comment by davidgraeff** [comment]
Done. As suggested I have introduced a `LedHandlerInterface`. I also merged `MilightBridgeDiscovery` and `MilightDiscover`, the separation is not necessary anymore since the clean up.

File: addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/LedHandlerInterface.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+@NonNullByDefault
+public interface LedHandlerInterface {
+    /**
+     *
+     * @param hue A value from 0 to 360
+     * @param saturation A saturation value. Can be -1 if not known
+     * @param brightness A brightness value. Can be -1 if not known
+     * @param state The changed values will be written back to the state
+     */
+    public abstract void setHSB(int hue, int saturation, int brightness, MilightThingState state);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/LedHandlerInterface.java]: For the interface you should preferably lose all the `public abstract`

**Comment by davidgraeff** [comment]
Squashed and removed "Public abstract" from the interface

**Comment by martinvw** [comment]
@davidgraeff please add the breaking changes warning to https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/update.lst

**Comment by martinvw** [comment]
@davidgraeff could you maybe update the title to something that suits the release notes?

**Pull request merged**